---
title: " Chapter II - Quality Control and metadata analysis "
output: html_notebook
---




# 3. Quality control before alingment


## 3.0 Getting the software

To perform a QC of the Fastq files we can use *FastQC*. In order to get this software we have several alternatives, and one option is to download it from [Babraham Bioinformatics](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). However, this might be time-consuming when we want to install different software to conduct our analysis. Instead, we can install software from [Conda](https://docs.conda.io/en/latest/), which is multi-platform a package repository that will enable us to quickly get the software and automatically install all their dependencies. 

We currently have conda installed on the machines. However, before start using it we need to make sure that some conda channels are added into our configuration. Add these channels by typing:

`conda config --add channels defaults`
`conda config --add channels conda-forge`
`conda config --add channels bioconda`


Then to install `fastqc` execute the following command.

`conda install fastqc`

To test if `fastqc` we can check `fastqc` manual by running.

`fastqc --help`

However, since `fasqc` is not going to be the only software we are going to use for the QC, we will create a new environment where we will install the tools we will use on this Chapter. To create a new environment called `QC` and install `fastqc` on it, we can run:

`conda create -n QC fastqc`

Then we can activate this new enviroment by:

`conda activate QC`

Now, check if `fastqc` was successfully installed. Does it work? To learn more about this, please take a quick read to this [blog post](https://john-quensen.com/tutorials/conda-and-virtual-environments/). 

**Exercise 3.0.1**

Open a new terminal, connect to the server and create a new screen session. Then, create a new environment called R, installing `rstudio` on it. To ensure that is the correct version, use `--channel r` which enable to use a conda channel that is specialized on R packages. This will take a while to resolve, but it corresponds to an environment that we will use on the last part of this chapter, so please keep an eye on it and introduce `y` when conda finish to gather all the dependencies. 

## 3.1 FASTQC


**Exercise 3.1.1**

Create a directory named `QC`. Read `fastqc` help manual to run it for `FASTQ/SRR7889585.fastq.gz` and store the results at the newly created folder (`QC`). To speed it up, use 8 threads (which it will allow our machines to use 8 processors). You will get an html file as output; use `scp` to copy it to your local machine. On a new terminal insert:

`scp course@IP_machine:/path/to/SRR7889585_fastqc.html .`

Now, open it with an internet browser (Chrome, Firefox, etc) and answer the following questions:

A) What is the read length?
B) Does the quality score very through the read length?
C) How is the data's quality?

## 3.2  Pre-processing

Given the quality of our experiment, we can make decisions about the following steps. If the data quality is not great, some pre-processing steps can be implemented to improve quality. Most of these steps can be conducted with [FASTX-Toolkit](http://hannonlab.cshl.edu/fastx_toolkit/commandline.html). 
The most common steps to pre-process FASTQ files are **adaptor removal** and **quality trimming**. Adaptors can be removed using `fastx_clipper`. In our case, our report does not show an evident adapter content and since we do not have the information about what adapters were used, we will skip this step. According to the report, the 3` ends of `
SRR7889585.fastq.gz` are not very low quality, but it worth to check if it can be improved. 

To install FASTX-Toolkit we are going to use `anaconda` to search for the right package. Please activate the `base` enviroment and then install `anaconda`. Once you have done this type:

`anaconda search -t conda FASTX | less `

You will get a list will all the channels where FASTX is available. To select the first option we can do:

`anaconda show BioBuilds/fastx-toolkit`

Take a look to the last conda command that its displayed. Activate the QC enviroment again and run that command. 


**Exercise 3.2.1**


A) The check the manual of fastq_quality_trimmer by running:

`fastq_quality_trimmer -h`

B) Generate an unziped copy of `SRR7889585.fastq.gz` by using:

`zcat SRR7889585.fastq.gz > SRR7889585.fastq`

Then find the right command to trim low quality sequences that are bellow phed score of 10 and only keep the sequences that are remain longer than 40 bp. 

C) Compare the number of lines between trimmed and untrimmed SRR7889585 files. Perform a QC report of the new trimmed file, and compare it with the original. Do you notice any improvement? 


## 3.3 Metadata analysis on R-studio


The fastq.gz files that we have at `/FASTQ` correspond to RNA-seq data extracted from D. melanogaster brain samples. The data corresponding to a full study published last year on [Cell Reports](https://www.ncbi.nlm.nih.gov//pubmed/30355484) and the full study was published in [SRA archive](https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP162335). 

**Exercise 3.3.1**

Check the [study runs](https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP162335) and compare with the file names from `FASTQ/` folder. Which tissues the samples were extracted from? Select the Runs that correspond to the files that we have and download the `RunInfo Table` corresponding to these tables and save it as `SraRunTable.txt` on the course folder. 

**Exercise 3.3.2**


Open a new terminal and connect to the server using `ssh`, but this time adding `-Y` to enable visualization of graphical interfaces. Then, activate the environment you created on Exercise 3.0.1 and enter the course folder. Run R-studio on the terminal by typing `rstudio`. An R-studio window should pop up. Install `ggplot2` and `data.table` packages by running the following line on the command line:

install.packages("ggplot2", "data.table")

This R-studio visualization is runnig in our server, so then we can open
`ChapterII_QC_and_metadata_analysis.Rmd` and continue following this chapter on R-studio! 

 that is located at the course folder. This is an R-notebook file, that has all the code used to generate this document. But, now when you open this file you will be able to run `chunks` of code interactively inside R-studio. Run the following chunk by clicking at the green play button.

- - - - - Continue reading inside the R-sudio - - - - - -

 This is an R-notebook file, that has all the code used to generate this document. But, now when you open this file you will be able to run `chunks` of code interactively inside R-studio. Run the following chunk by clicking at the green play button.

```{r}

library(data.table)

metadata <- fread("./SraRunTable.txt") #Please be sure to place the file under ~/RNA_seq_course2020

metadata[ source_name=="central neurons"  , sum(AvgSpotLen) ]
```


By running this code, we imported the metadata as a data.table object, which corresponds to a very convenient structure to manipulate data in R. For more information visit this [link](https://www.datacamp.com/community/tutorials/data-table-cheat-sheet). We can now easily manipulate the data! For example, we can filter the metadata rows, so we only see the ones corresponding to "central neurons"


```{r}
metadata[source_name=="central neurons" , ]
```


We can also get explore if we have a lowly sequenced sample filtering by the `MBases` values

```{r}

metadata[Bases < 100000000 , ]
```


Or get the average Mega bases across this sample

```{r}

metadata[  , mean(Bases), by=source_name ]

```


To explore this visually, we can plot with ggplot2. This is one of the most popular packages to visualize data. A basic code to visualize the Mbases distribution can be run as:

```{r}
library(ggplot2)

ggplot( metadata) +  # Frist layer - Data input
  geom_bar(aes(x=Run, y=Bases), stat = "identity")   # second layer - type of plot and axis

```


We can rotate the text to make it more visible

```{r}

ggplot(data=metadata) +
  geom_bar(aes(x=Run, y=Bases, ), stat = "identity") + 
  theme(axis.text.x = element_text(angle = 45))  # Third layer -  visual configuration

```


inside `aes( )` are the variables and we can also define a variable as the colour:


```{r}
ggplot(data=metadata) +
  geom_bar(aes(x=Run, y=Bases, colour=source_name), stat = "identity") + 
  theme(axis.text.x = element_text(angle = 45))
```

But for this particular type of graph, filling the bar with one colour is more suitable

```{r}
ggplot(data=metadata) +
  geom_bar(aes(x=Run, y=Bases, fill=source_name), stat = "identity") + 
  theme(axis.text.x = element_text(angle = 45))
```


From this plot, we can clearly see that one of the samples that were taken from central neurons is much smaller than the rest. This factor needs to be considered at the time we analyze the samples, as this poorly sequenced sample might have lead to less accurate mRNA quantification and it might need to be removed for quantitative analyses. 

```{r}
metadata
```


